### PR TITLE
Issue #481: Reduce db calls

### PR DESCRIPTION
**Pull request by petersistrom**
#481

### CODE REVIEW

File: classes/dataflow.php [hunk 1]
```diff
@@ -550,8 +560,8 @@ public function get_dotscript(): string {
         ];
         foreach ($edges as $edge) {
             [$srcid, $destid] = $edge;
-            $srcstep = new step($srcid);
-            $deststep = new step($destid);
+            $srcstep = $steparray[$srcid];
```
Comment (brendanheywood) [inline classes/dataflow.php:563]: This bit make sense so keep this

File: classes/steps_table.php [hunk 2]
```diff
@@ -79,6 +79,7 @@ public function make_columns(): void {
      */
     public function col_name(\stdClass $record): string {
         $step = new step($record->id);
+        $step->states->valid = true;
```
Comment (brendanheywood) [inline classes/steps_table.php:82]: this feels out of place, how does this one help?

File: visual.php [hunk 3]
```diff
@@ -38,6 +38,7 @@
 
 // Generate DOT script based on the configured dataflow.
 $dataflow = new dataflow($id);
+$dataflow->validate_dataflow();
```
Comment (brendanheywood) [inline visual.php:41]: what side effects does this have here which help?

File: classes/steps_table.php (hunk 2 above)
Comment (petersistrom) [reply classes/steps_table.php:82]: Changing this line https://github.com/catalyst/moodle-tool_dataflows/pull/529/files#diff-642f9f96d3487795294cba491b20d8119c2d15032bb22abeb72c475a88191cf5L859 made all the steps in the table false but it definitely needs some tweaking

File: visual.php (hunk 3 above)
Comment (petersistrom) [reply visual.php:41]: Sets the valid state for each step, so each step doesn't have to be validated individually again

File: classes/steps_table.php (hunk 2 above)
Comment (brendanheywood) [reply classes/steps_table.php:82]: ok not sure this is the right approach so lets land the other bits which are fine and the changes yet to come for the persistent

File: classes/step.php [hunk 4]
```diff
@@ -449,9 +460,12 @@ public function dependents() {
              LEFT JOIN {tool_dataflows_steps} step ON sd.stepid = step.id
                  WHERE sd.dependson = :stepid";
 
-        $deps = $DB->get_records_sql($sql, [
-            'stepid' => $this->id,
-        ]);
+        if (empty($this->dependents) || $reload) {
```
Comment (brendanheywood) [inline classes/step.php:463]: should this be isset rather than empty and the attribute should default to null rather than an empty array?

File: classes/step.php [hunk 5]
```diff
@@ -635,6 +649,7 @@ public function validate_step() {
                 $errors = array_merge($errors, $extravalidation);
             }
         }
+
```
Comment (brendanheywood) [inline classes/step.php:652]: try to avoid unrelated changes

File: classes/step.php [hunk 6]
```diff
@@ -849,8 +864,8 @@ public function get_dotscript($contentonly = false): string {
         } else {
             // Invalid step type styles.
             $stepstyles = [
-               'fillcolor' => '#ca3120',
-               'fontcolor' => '#ffffff',
+                'fillcolor' => '#ca3120',
```
Comment (brendanheywood) [inline classes/step.php:867]: things like this are good to cleanup, but better to be in a different pr or commit

File: classes/step.php [hunk 7]
```diff
@@ -427,19 +434,23 @@ public function dependants(): array {
...
-        $deps = $DB->get_records_sql($sql, [
-            'dependson' => $this->id,
-        ]);
+        if (!isset($this->dependants) || $reload) {
+            $deps = $DB->get_records_sql($sql, [
+                'dependson' => $this->id,
+            ]);
+            $this->dependants = $deps;
+        }
         return $deps ?? [];
```
Comment (brendanheywood) [inline classes/step.php:443]: this should always return $this->dependants;

File: classes/step.php [hunk 8]
```diff
@@ -449,10 +460,13 @@ public function dependents() {
...
-            'stepid' => $this->id,
-        ]);
-        return $deps;
+        if (!isset($this->dependents) || $reload) {
+            $deps = $DB->get_records_sql($sql, [
+                'stepid' => $this->id,
+            ]);
+            $this->dependents = $deps;
+        }
+        return $deps ?? [];
```
Comment (brendanheywood) [inline classes/step.php:469]: same

File: classes/dataflow.php [hunk 9]
```diff
@@ -533,9 +533,14 @@ public function get_dotscript(): string {
 
         // Fetch the dot script node from each step to construct them.
         $steps = $this->steps;
+
         $nodes = [];
         $contentonly = true;
+        $steparray = [];
+        $oputlabels = [];
```
Comment (keevan) [inline classes/dataflow.php:540]:
```suggestion
        $outputlabels = [];
```
Did you mean this?

File: classes/step.php [hunk 10]
```diff
@@ -49,6 +49,12 @@ class step extends persistent {
     /** @var array $dependson */
     private $dependson = [];
 
+    /** @var array $dependson */
+    private $dependants = null;
+
+    /** @var array $dependson */
+    private $dependents = null;
```
Comment (keevan) [inline classes/step.php:56]: The variable it's referencing is incorrect (copy + paste but no edit).

Are you able to add a brief description to the comment, because it adds no value currently (aside from type specification)? i.e. it would be good to know what the variable means and is for.

**Comment by jaypha** [comment]
The image in the dataflows manager screen is not updating in line with changes to the dataflow.
For example
Screenshot from 2022-09-06 17-17-07
When it should be.
Screenshot from 2022-09-06 17-17-45

**Comment by brendanheywood** [comment]
> The image in the dataflows manager screen is not updating in line with changes to the dataflow. For example 

This hints that there is probably deeper issues which the cache not being invalidated and there is likely a range of related bugs with the same root cause

**Comment by petersistrom** [comment]
I'm not sure this is related to this PR. Seems to be happening on STABLE as well. A hard refresh of the browser will load the correct preview image.

**Comment by jaypha** [comment]
I am seeing this too.
